### PR TITLE
Early exit auth check on lease puts

### PR DESCRIPTION
**Pull request by tjungblu**
Mitigates etcd-io#15993 by not checking each key individually for permission when auth is entirely disabled or admin user is calling the method.

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Review by jmhbnz — COMMENTED** [review]
Hey @tjungblu nice work looking into this!

Any chance you could replicate the metrics in the linked issue and post here so we could compare before and after?

File: server/etcdserver/apply/apply_auth.go
Comment (serathius) [inline server/etcdserver/apply/apply_auth.go:130]: Do we have a test covering this case?

File: server/etcdserver/apply/apply_auth.go
Comment (serathius) [inline server/etcdserver/apply/apply_auth.go:129]: It's not obvious from proposed check that it covers case that auth is disabled. From code design it seems strange  for me that `authApplierV3` will be used at all if auth is disabled.

Comment (tjungblu) [reply server/etcdserver/apply/apply_auth.go:129]: I was surprised too, it even goes through an rw mutex everytime it checks...

https://github.com/etcd-io/etcd/issues/15993#issuecomment-1573550357

I reckon we should discuss a bigger refactoring for the auth system, potentially?

File: server/etcdserver/apply/apply_auth.go
Comment (tjungblu) [reply server/etcdserver/apply/apply_auth.go:130]: Neither apply.go, nor apply_auth.go have unit test coverage-  there are no unit tests as far as I can tell.

File: server/etcdserver/apply/apply_auth.go
Comment (serathius) [reply server/etcdserver/apply/apply_auth.go:129]: Maybe, depends on whether we want to backport this fix. This seems like a trivial but important change. Appliers changed a lot between v3.5 and main.

Comment (ahrtr) [reply server/etcdserver/apply/apply_auth.go:129]: > From code design it seems strange for me that authApplierV3 will be used at all if auth is disabled.

> I reckon we should discuss a bigger refactoring for the auth system, potentially?

Because the existing chain of applier is static. If we want to get rid of authApplier completely when auth isn't enabled, then we need to make the chain dynamic. We don't have to do it, please feel free to evaluate the effort and impact separately if anyone is interested.

> It's not obvious from proposed check that it covers case that auth is disabled.

It's a common "issue" in all existing `(*authStore) IsXXXPermitted(...) error` methods, when auth isn't enabled, they return nil.

> depends on whether we want to backport this fix

I think we need to backport the fix.

**Comment by ahrtr** [comment]
The PR looks good to me. @tjungblu did you compare the performance before and after applying this PR?

**Comment by tjungblu** [comment]
running this entirely local with the test supplied from the issue, I can see the p99 latency to be consistently good:

image

here's the same result without the patch right afterwards in comparison:

image

The p99 looks very low in absolute by prometheus because of the scraping interval and one minute resolution, it actually degrades almost immediately into 100ms+ latency after 10k entries:

> MAX latency: 103.208013ms, entries: 11405 

which is basically a few seconds into the unit test. 


--- 

Just for completeness with a comparison of the same metric used in the issue report also stays flat over the whole execution time:
image

Even though I'm not entirely sure whether it makes sense to sum over the rate duration sum here.

File: server/etcdserver/apply/apply_auth.go
Comment (tjungblu) [reply server/etcdserver/apply/apply_auth.go:129]: > we need to make the chain dynamic.  

I haven't researched too deeply yet, but the whole system is already dynamic because the whole auth goes through raft (even enable/disable). I would've expected the enable/disable methods to be a configuration for startup, not a dynamic apply request. Hence the static implementation, I guess.

I'll open a discussion issue up as a feature request, will do some benchmarks in the meantime around the impact of either approach.

Comment (ahrtr) [reply server/etcdserver/apply/apply_auth.go:129]: The link you provide just replaces the whole applier chain with `applierV3Capped` or `applierV3Corrupt`. But you can't change (swap out) part of the chain: `authApplier -> quotaApplier -> backendApplier`. 

It's similar to `string` in golang, it's immutable. You can't modify part of a string; instead, you can only replace the whole string.

Just as I mentioned previously, we don't have to make the applier chain dynamic. But of course, it's open to discussion in a separate session.

Comment (tjungblu) [reply server/etcdserver/apply/apply_auth.go]: updated

Comment (tjungblu) [reply server/etcdserver/apply/apply_auth.go]: No doubt on the immutability of the chain, but the pointer to the applier struct needs to be guarded by a mutex when you swap it out from another goroutine?

https://github.com/etcd-io/etcd/blob/release-3.5/server/etcdserver/server.go#L252

let me try to tease the race detector on this case more specifically. Not that it matters much, if you're corrupted/OOS it's game over anyway, it just tickled my inner race detector.

**Review by ahrtr — APPROVED** [review]
LGTM

Thanks @tjungblu 

It's a straightforward change, and also a low-hang fruit. 

Could you please double check also backport this PR to 3.5 and 3.4 if needed? @tjungblu  Of course, waiting for other maintainers approve and merge this PR.

**Comment by tjungblu** [comment]
@marseel I assume you would need this at the very least in the next 3.5.x release?

**Comment by marseel** [comment]
3.5.x release would be great.

**Comment by tjungblu** [comment]
cool, then I'll get to write a backport. Thanks everyone!

**Review by jmhbnz — APPROVED** [review]

**Comment by serathius** [comment]
One think before merging. @tjungblu I know we don't have tests, however I would feel much safer for backports if we could add some even trivial test.

**Comment by tjungblu** [comment]
Totally makes sense, let me try to add one right here as a separate commit. I'll cherry-pick it down for the backport PRs then.

**Review by serathius — CHANGES_REQUESTED** [review]
Can we add a trivial test to make backports safer?

File: server/etcdserver/apply/apply_auth.go
Comment (serathius) [reply server/etcdserver/apply/apply_auth.go:131]: My suggestion would be to add `ArePutsPermitted` method that will for the range and add tests for it.

**Comment by tjungblu** [comment]
@serathius can we move this forward? Is the testcase sufficient for you or too much refactoring already?

File: server/auth/store_mock.go
Comment (serathius) [inline server/auth/store_mock.go:14]: Don't move test file to normal file. Don't share mocks outside of package.

File: server/auth/nop.go
Comment (serathius) [inline server/auth/nop.go:21]: Don't expose internal testing code from package! Use proper TokenProvider in external tests.

Comment (tjungblu) [reply server/auth/nop.go]: reverted

File: server/etcdserver/apply/apply_auth_test.go
Comment (serathius) [inline server/etcdserver/apply/apply_auth_test.go:55]:
```suggestion
	assert.NoError(t, aa.checkLeasePutsKeys(lease.NewLease(lease.LeaseID(1), 3600)), "auth is disabled, should allow puts")
```

**Comment by serathius** [comment]
Overall looks very good! Tests you added are awesome! Left some small comments.

File: server/etcdserver/apply/apply_auth_test.go
Comment (tjungblu) [reply server/etcdserver/apply/apply_auth_test.go:52]: replaced with bcrypt.MinCost

**Comment by tjungblu** [comment]
fixed the goimport manually, I reckon `make fix` doesn't actually do anything?

**Review by serathius — APPROVED** [review]

**Pull request merged**
